run-csharp

```diff
@@ -2,10 +2,71 @@
 
 . ./config
 
+CSHARP_TEST_FORMATS_DIR=$FORMATS_COMPILED_DIR/csharp
 CSHARP_SPEC_DIR=spec/csharp
+CSHARP_TEST_OUT_DIR="$TEST_OUT_DIR"/csharp
 
-# Compile everything
-xbuild "$CSHARP_SPEC_DIR/kaitai_struct_csharp_tests.sln"
+rm -rf "$CSHARP_TEST_OUT_DIR"
+mkdir -p "$CSHARP_TEST_OUT_DIR"
+
+# Try and build the project first - if there's no compile errors, we can continue straight away
+echo "run-csharp: building project"
+xbuild "$CSHARP_SPEC_DIR/kaitai_struct_csharp_tests.sln" >> "$CSHARP_TEST_OUT_DIR/_compile.log" || {
+
+    # Some formats failed to build - find the failing files and erase them
+	echo "run-csharp: some formats failed to compile, checking each file individually"
+    
+    # Find the invalid files and wipe them - we do this by scanning the error list from the compiler
+    mcs \
+        -t:library \
+        -fullpaths \
+        -o:$CSHARP_TEST_OUT_DIR/_cs_compile_temp.dll \
+        -r:"$CSHARP_SPEC_DIR/packages/NUnit.3.4.1/lib/net45/nunit.framework.dll" \
+        "$CSHARP_RUNTIME_DIR/*.cs" \
+        "$CSHARP_SPEC_DIR/kaitai_struct_csharp_tests/CommonSpec.cs" \
+        "$CSHARP_TEST_FORMATS_DIR"/*.cs \
+            2>&1 >> "$CSHARP_TEST_OUT_DIR/_compile.log" | \
+            while read -r line ; do
+                # e.g. path/to/file.cs(1): Unexpected symbol
+                # Grab everything before the first '(' character
+                FILENAME=$(echo $line | cut -d '(' -f 1)
+                if [ -s $FILENAME ]
+                then
+                    NAME=$(basename "$FILENAME")
+                    echo "$NAME failed to compile, removing"
+                    > $FILENAME
+                fi
+            done
```

Comment:
You don't actually remove the files, you replace them with empty ones, do you? I'd rather kept it a bit simpler and more readable:

```shell
COMPILE_LOG1="$CSHARP_TEST_OUT_DIR/_compile1.log"

mcs blah blah 2>&1 >"$COMPILE_LOG1" || {
    fgrep '(' "$COMPILE_LOG1" | cut -d'(' -f1 >"$FAILED1"

    echo "Failed to compile, removing broken files:"
    cat "$FAILED1"

    cat "$FAILED1" | xargs truncate -s0
}
```

Even so, I don't really see much point in listing emptied files other than debugging the script - for informative purposes, they all got listed above anyway. This might make it as short as:
```shell
COMPILE_LOG1="$CSHARP_TEST_OUT_DIR/_compile1.log"

mcs blah blah 2>&1 >"$COMPILE_LOG1" ||
    fgrep '(' "$COMPILE_LOG1" | cut -d'(' -f1 | xargs truncate -s0
```